Math/src/main/java/org/tribuo/math/neighbour/kdtree/DimensionNode.java

```diff
@@ -0,0 +1,168 @@
+     * @param record The point.
+     * @param distanceType The distance function.
+     */
+    public DimensionNode(int dimension, IntAndVector record, DistanceType distanceType) {
+        this.dimension = dimension;
+        this.record = record;
+        this.maxD = record.vector.size();
+        this.distanceType = distanceType;
+        this.coord = record.vector.get(dimension-1);
+    }
+
+    /**
+     * Return the node below this one in the k-d tree.
+     * @return The node below this one.
+     */
+    public DimensionNode getBelow() {
+        return below;
+    }
+
+    /**
+     * Return the node above this one in the k-d tree.
+     * @return The node below this one.
+     */
+    public DimensionNode getAbove() {
+        return above;
+    }
+
+    /**
+     * Set the node below this one in the k-d tree.
+     *
+     * @param node The node being added to the tree.
+     */
+    public void setBelow(DimensionNode node) {
+        if (node == null) {
+            this.below = null;
+        }
+        // Ensure the dimension of node being added is only 1 greater, or is the dimension wrap around case.
+        else if ((this.dimension + 1 == node.dimension) || (this.dimension == maxD && node.dimension == 1)) {
+            this.below = node;
+        }
+        else {
+            throw new IllegalArgumentException("Setting the below/left node failed because the dimensions are incorrect.");
+        }
+    }
+
+    /**
+     * Set the node above this one in the k-d tree.
+     *
+     * @param node The node being added to the tree.
+     */
+    public void setAbove(DimensionNode node) {
+        if (node == null) {
+            this.above = null;
+        }
+        // Ensure the dimension of node being added is only 1 greater, or is the dimension wrap around case.
+        else if ((this.dimension + 1 == node.dimension) || (this.dimension == maxD && node.dimension == 1)) {
+            this.above = node;
+        }
+        else {
+            throw new IllegalArgumentException("Setting the above/right node failed because the dimensions are incorrect.");
+        }
+    }
+
+    /**
+     * Tests if the provided point is strictly below the line this node and its dimension represent.
+     *
+     * @param point The point being checked.
+     * @return true when the provided point is below this node.
+     */
+    boolean isBelow(SGDVector point) {
+        return point.get(dimension - 1) < coord;
+    }
+
+    /**
+     * Traverse the (sub)tree rooted at this node to see which of its descendants are closest to the provided point.
+     *
+     * @param point The target point.
+     * @param queue The priority queue used to maintain the k nearest neighbours.
+     * @param isInitializing A flag which indicates if this method is being called during query initialization.
+     */
+    void nearest(SGDVector point, DistanceRecordBoundedMinHeap queue, boolean isInitializing) {
```

Comment:
How unbalanced can this tree get if we adversarially prepare a dataset? Can it pop the stack with the recursion in this call?

Reply:
This algorithm always constructs a balanced tree, but I suppose there would be some manually generated datasets that could be provided that cause the algorithm to perform poorly. With recursion, there is always the risk that a tree constructed from a huge dataset could overflow the stack, but I would suspect we may also face memory issues in these scenarios. It sounds like you have more experience with this type of issue, so I'd be interested in hearing more.

Reply:
So I agree that it is picking the median, but if the array is `[0,1,1,1,1,1,1,1]` then the invariants say that it should split that into `[0]` and `[1,1,1,1,1,1,1]`, so I wonder what happens if you have a single data point that's all zeros and then every other data point is the same and all ones. We can decide that adversarial datasets are not our problem, but I'd like to have a rough idea how adversarial it has to be before it goes pop.

Reply:
I may not be understanding the issue here. If I use this code to create a tree from the points: (0,0,0,0), and 9 points like (1,1,1,1), I still get a balanced tree. The root node, will be (1,1,1,1). Traversing the tree following the below nodes, arrives at the leaf point (0,0,0,0). All the other nodes in the tree are (1,1,1,1). I have added a test to the performance test branch I'm maintaining where this scenario could be executed in the debugger: [adversarial test](https://github.com/geoffreydstewart/tribuo/blob/kd-tree-perf-tests/Clustering/Core/src/test/java/org/tribuo/neighbours/TestBreakdownKDTreePerf.java#L169)

Reply:
Ok. I had a look through the test, when I run it the trees are balanced, but points with value equal to the split point appear on the left. I thought the invariant was that that couldn't happen? I've left some comments on the new PR in relevant places.